api/metric/api.go

```diff
@@ -21,62 +23,231 @@ import (
 	"go.opentelemetry.io/api/unit"
 )
 
-type MetricType int
+// Kind categorizes different kinds of metric.
+type Kind int
 
 const (
-	Invalid    MetricType = iota
-	Gauge                 // Supports Set()
-	Cumulative            // Supports Inc()
+	Invalid     Kind = iota
+	CounterKind      // Supports Add()
+	GaugeKind        // Supports Set()
+	MeasureKind      // Supports Record()
+	ObserverKind
 )
 
-type Meter interface {
-	// TODO more Metric types
-	GetFloat64Gauge(ctx context.Context, gauge *Float64GaugeHandle, labels ...core.KeyValue) Float64Gauge
+// Recorder is the implementation-level interface to Set/Add/Record individual metrics.
+type Handle interface {
+	// Record allows the SDK to observe a single metric event
+	RecordOne(ctx context.Context, value MeasurementValue)
 }
 
-type Float64Gauge interface {
-	Set(ctx context.Context, value float64, labels ...core.KeyValue)
+// LabelSet represents a []core.KeyValue for use as pre-defined labels
+// in the metrics API.
+//
+// TODO this belongs outside the metrics API, in some sense, but that
+// might create a dependency.  Putting this here means we can't re-use
+// a LabelSet between metrics and tracing, even when they are the same
+// SDK.
+type LabelSet interface {
+	Meter() Meter
 }
 
-type Handle struct {
-	Name        string
-	Description string
-	Unit        unit.Unit
+// ObserverCallback defines a type of the callback SDK will call for
+// the registered observers.
+type ObserverCallback func(Meter, Observer) (LabelSet, MeasurementValue)
```

Comment:
I believe the intention was to return multiple `LabelSet, MeasurementValue` pairs from the callback. Potentially a more-performant approach would be a `func(Meter, Observer, func(LabelSet, MeasurementValue)` that passes a callback to the callback, allowing it to emit these pairs one at a time.

Reply:
Ah, will fix it then. Would it make sense to make it possible to provide both ways (return a map or repeatedly call the callback with the pair)?

Anyway, what's the use-case for returning multiple pairs from a single observer? Returning past values (that would likely need some timestamp overriding and whatnot)?

Reply:
Returning a map would be a PITA for the case of a specialized (typed) callback for either Int64Observer or Float64Observer - we would need to convert `map[metric.LabelSet]int64` to `map[metric.LabelSet]MeasurementValue`. I went with the callback in callback thing for now.

Reply:
I'll think about this more. Take a look at #172 -- the whole int64 vs float64 thing is almost out-of-hand.